symengine/fields.cpp

```diff
@@ -0,0 +1,241 @@
+}
+
+std::size_t GaloisField::__hash__() const
+{
+    std::hash<std::string> hash_string;
+    std::size_t seed = GALOISFIELD;
+
+    seed += hash_string(this->var_->get_name());
+    for (const auto &it : poly_.dict_) {
+        std::size_t temp = GALOISFIELD;
+        hash_combine<unsigned int>(temp, it.first);
+        hash_combine<long long int>(temp, mp_get_si(it.second));
+        seed += temp;
+    }
+    return seed;
+}
+
+int GaloisField::compare(const Basic &o) const
+{
+    const GaloisField &s = static_cast<const GaloisField &>(o);
+
+    if (poly_.size() != s.poly_.size())
+        return (poly_.size() < s.poly_.size()) ? -1 : 1;
+
+    int cmp = var_->compare(*s.var_);
+    if (cmp != 0)
+        return cmp;
+    return 0; // TODO
+}
+
+
+RCP<const GaloisField> GaloisField::from_dict(const RCP<const Symbol> &var,
+                                        GaloisFieldDict &&d)
+{
+    return make_rcp<const GaloisField>(var, std::move(d));
+}
+
+RCP<const GaloisField> GaloisField::from_vec(const RCP<const Symbol> &var,
+                                       const std::vector<integer_class> &v,
+                                       const integer_class &modulo)
+{
+    return make_rcp<const GaloisField>(var, GaloisFieldDict::from_vec(v, modulo));
+}
+
+vec_basic GaloisField::get_args() const
+{
+    vec_basic args;
+    for (const auto &p : poly_.dict_) {
+        if (p.first == 0) {
+            args.push_back(integer(p.second));
+        } else if (p.first == 1) {
+            if (p.second == 1) {
+                args.push_back(var_);
+            } else {
+                args.push_back(
+                    Mul::from_dict(integer(p.second), {{var_, one}}));
+            }
+        } else {
+            if (p.second == 1) {
+                args.push_back(pow(var_, integer(p.first)));
+            } else {
+                args.push_back(Mul::from_dict(integer(p.second),
+                                              {{var_, integer(p.first)}}));
+            }
+        }
+    }
+    if (poly_.dict_.empty())
+        args.push_back(zero);
+    return args;
+}
+
+void GaloisFieldDict::gf_div(const GaloisFieldDict &o,
+                         const Ptr<GaloisFieldDict> &quo,
+                         const Ptr<GaloisFieldDict> &rem) const
+{
+    SYMENGINE_ASSERT(modulo_ == o.modulo_);
+    if (o.dict_.empty())
+        throw std::runtime_error("ZeroDivisionError");
+    map_uint_mpz dict_divisor = o.dict_;
+    map_uint_mpz dict_out;
+    size_t deg_dividend = dict_.empty() ? 0 : dict_.rbegin()->first;
+    size_t deg_divisor = dict_divisor.rbegin()->first;
+    if (deg_dividend < deg_divisor) {
+        *quo = GaloisFieldDict(dict_out, modulo_);
+        *rem = GaloisFieldDict(dict_, modulo_);
+    } else {
+        dict_out = dict_;
+        integer_class inv;
+        mp_invert(inv, dict_divisor.rbegin()->second, modulo_);
+        integer_class coeff;
+        for (size_t it = deg_dividend; it != size_t(-1); it--) {
```

Comment:
`size_t` is an unsigned data type. This will make `size_t(-1)` a large positive number. Use `long` here